esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
     out.clear().append(finger-start,start);
 }
 
+void EsdlServiceImpl::adjustTargetConfig(Owned<IPTree>& tgtcfg) const
+{
+    static const VStringBuffer gwLocalSecretXPath("Gateways/Gateway[@url='%s*']", gwLocalSecretPrefix);
+
+    if (!tgtcfg)
+        return;
+
+    if (tgtcfg->hasProp(gwLocalSecretXPath))
+    {
+        tgtcfg.setown(createPTreeFromIPT(tgtcfg));
+        const char* qt = tgtcfg->queryProp("@querytype");
+        bool published = (qt && (strieq(qt, "roxie") || strieq(qt, "wsecl")));
+        const char* permission = (published ? "allowPublishedGatewayUsage" : "allowUnpublishedGatewayUsage");
+        Owned<IPTreeIterator> gwsToResolve(tgtcfg->getElements(gwLocalSecretXPath));
+        ForEach(*gwsToResolve)
+            resolveGatewayLocalSecret(gwsToResolve->query(), permission);
+    }
+}
+
+void EsdlServiceImpl::resolveGatewayLocalSecret(IPTree& gateway, const char* permission) const
+{
+    const char* name = gateway.queryProp("@name");
+    StringBuffer url(gateway.queryProp("@url"));
+    if (url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: missing url property", name);
+    if (strncmp(url, gwLocalSecretPrefix, gwLocalSecretPrefixLength))
+        throw makeStringExceptionV(-1, "gateway %s: expected '%s...'; got '%s'", name, gwLocalSecretPrefix, url.str());
+    const char* identification = url.str() + gwLocalSecretPrefixLength;
+    StringArray tokens;
+    tokens.appendList(identification, ":", true);
+    Owned<IPTree> secret;
+    switch (tokens.ordinality())
+    {
+    case 1:
+        secret.setown(getSecret("esp", tokens.item(0)));
+        break;
+    case 2:
+        secret.setown(getVaultSecret("esp", tokens.item(0), tokens.item(1)));
+        break;
+    default:
+        throw makeStringExceptionV(-1, "gateway %s: '%s' is not of the form \"[ vault-id ':' ] secret-name\"", name, identification);
+    }
+    if (!secret)
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not identify an 'esp' category secret", name, identification);
+    if (!isEmptyString(permission) && !secret->getPropBool(permission))
+        throw makeStringExceptionV(-1, "gateway %s: '%s' does not grant '%s' permission", name, identification, permission);
+    if (!secret->getProp("url", url.clear()) || url.isEmpty())
```

Comment:
The url will always get cleared to prevent getProp from appending the secret's defined URL to the configurations secret identification . The second clause is a sanity check that the secret didn't define an empty URL. I think the logic is as intended.

Reply:
Ok I guess I misunderstood. I thought there could be a valid case here where it wasn't required to resolve the secret.